File: app/src/androidTest/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatalist/WeatherDataListActivityTest.java
package pl.mosenko.sunnypodlaskie.mvp.weatherdatalist;

import android.app.Activity;
import android.support.test.InstrumentationRegistry;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.support.test.runner.lifecycle.ActivityLifecycleMonitorRegistry;
import android.support.test.runner.lifecycle.Stage;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collection;

import pl.mosenko.sunnypodlaskie.R;
import pl.mosenko.sunnypodlaskie.mvp.setting.SettingsActivity;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static org.junit.Assert.assertTrue;

/**
 * Created by syk on 10.06.17.
 */

@RunWith(AndroidJUnit4.class)
public class WeatherDataListActivityTest {

    @Rule
    public ActivityTestRule<WeatherDataListActivity> listActivityActivityTestRule =
            new ActivityTestRule<WeatherDataListActivity>(WeatherDataListActivity.class);

    @Test
    public void clickOnSetting_shouldStartSettingActivity() {
        onView(withId(R.id.WeatherDataListActivity_action_settings)).perform(click());
        Activity activity = getActivityInstance();
        assertTrue(activity instanceof SettingsActivity);
    }

    public Activity getActivityInstance() {
        final Activity[] activity = new Activity[1];

        InstrumentationRegistry.getInstrumentation().runOnMainSync(() -> {
            Activity currentActivity = null;
            Collection resumedActivities = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(Stage.RESUMED);
            if (resumedActivities.iterator().hasNext()) {
                currentActivity = (Activity) resumedActivities.iterator().next();
                activity[0] = currentActivity;
            }
        });
        return activity[0];
    }
}

File: app/src/androidTest/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatalist/WeatherDataListActivityTest.kt
package pl.mosenko.sunnypodlaskie.mvp.weatherdatalist

import android.app.Activity
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.rule.ActivityTestRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry
import androidx.test.runner.lifecycle.Stage
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import pl.mosenko.sunnypodlaskie.R
import pl.mosenko.sunnypodlaskie.mvp.setting.SettingsActivity

/**
 * Created by syk on 10.06.17.
 */
@RunWith(AndroidJUnit4::class)
class WeatherDataListActivityTest {
    @Rule
    var listActivityActivityTestRule: ActivityTestRule<WeatherDataListActivity?>? = ActivityTestRule(WeatherDataListActivity::class.java)
    @Test
    fun clickOnSetting_shouldStartSettingActivity() {
        Espresso.onView(ViewMatchers.withId(R.id.WeatherDataListActivity_action_settings)).perform(ViewActions.click())
        val activity = getActivityInstance()
        Assert.assertTrue(activity is SettingsActivity)
    }

    fun getActivityInstance(): Activity? {
        val activity = arrayOfNulls<Activity?>(1)
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            var currentActivity: Activity? = null
            val resumedActivities: MutableCollection<*>? = ActivityLifecycleMonitorRegistry.getInstance().getActivitiesInStage(Stage.RESUMED)
            if (resumedActivities.iterator().hasNext()) {
                currentActivity = resumedActivities.iterator().next() as Activity?
                activity[0] = currentActivity
            }
        }
        return activity[0]
    }
}
